In [45]:
import pandas as pd
import numpy as np



In [46]:
#print("Jupyter is working")df = pd.read_csv("../data/raw/ElectronicsProductsPricingData.csv")

df = pd.read_csv("../data/raw/ElectronicsProductsPricingData.csv")
df.head()

,id,prices.amountMax,prices.amountMin,prices.availability,prices.condition,prices.currency,prices.dateSeen,prices.isSale,prices.merchant,prices.shipping,...,name,primaryCategories,sourceURLs,upc,weight,Unnamed: 26,Unnamed: 27,Unnamed: 28,Unnamed: 29,Unnamed: 30
0,AVphzgbJLJeJML43fA0o,104.99,104.99,Yes,New,USD,"2017-03-30T06:00:00Z,2017-03-10T22:00:00Z,2017...",False,Bestbuy.com,NaN,...,Sanus VLF410B1 10-Inch Super Slim Full-Motion ...,Electronics,https://www.amazon.com/Sanus-VLF410B1-10-Inch-...,7.94E+11,32.8 pounds,NaN,NaN,NaN,NaN,NaN
1,AVpgMuGwLJeJML43KY_c,69.00,64.99,In Stock,New,USD,2017-12-14T06:00:00Z,True,Walmart.com,Expedited,...,Boytone - 2500W 2.1-Ch. Home Theater System - ...,Electronics,http://reviews.bestbuy.com/3545/4784804/review...,6.42E+11,14 pounds,NaN,NaN,NaN,NaN,NaN
2,AVpgMuGwLJeJML43KY_c,69.00,69.00,In Stock,New,USD,2017-09-08T05:00:00Z,False,Walmart.com,Expedited,...,Boytone - 2500W 2.1-Ch. Home Theater System - ...,Electronics,http://reviews.bestbuy.com/3545/4784804/review...,6.42E+11,14 pounds,NaN,NaN,NaN,NaN,NaN
3,AVpgMuGwLJeJML43KY_c,69.99,69.99,Yes,New,USD,2017-10-10T05:00:00Z,False,Bestbuy.com,NaN,...,Boytone - 2500W 2.1-Ch. Home Theater System - ...,Electronics,http://reviews.bestbuy.com/3545/4784804/review...,6.42E+11,14 pounds,NaN,NaN,NaN,NaN,NaN
4,AVpgMuGwLJeJML43KY_c,66.99,66.99,Yes,New,USD,2017-08-28T07:00:00Z,False,Bestbuy.com,NaN,...,Boytone - 2500W 2.1-Ch. Home Theater System - ...,Electronics,http://reviews.bestbuy.com/3545/4784804/review...,6.42E+11,14 pounds,NaN,NaN,NaN,NaN,NaN


In [47]:
df.shape

(7249, 31)

In [48]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 7249 entries, 0 to 7248
Data columns (total 31 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   7249 non-null   str    
 1   prices.amountMax     7249 non-null   float64
 2   prices.amountMin     7249 non-null   float64
 3   prices.availability  7249 non-null   str    
 4   prices.condition     7249 non-null   str    
 5   prices.currency      7249 non-null   str    
 6   prices.dateSeen      7249 non-null   str    
 7   prices.isSale        7249 non-null   bool   
 8   prices.merchant      7249 non-null   str    
 9   prices.shipping      4277 non-null   str    
 10  prices.sourceURLs    7249 non-null   str    
 11  asins                7249 non-null   str    
 12  brand                7249 non-null   str    
 13  categories           7249 non-null   str    
 14  dateAdded            7249 non-null   str    
 15  dateUpdated          7249 non-null   str    
 16 

## 2. Data Quality Assessment

Before cleaning the dataset, I examined missing values, duplicate records, data types, and inconsistent values to determine which cleaning operations were necessary.

In [49]:
missing_values = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percent": (df.isnull().sum() / len(df) * 100).round(2)
})

missing_values = missing_values[
    missing_values["missing_count"] > 0
].sort_values("missing_percent", ascending=False)

missing_values

,missing_count,missing_percent
Unnamed: 28,7243,99.92
Unnamed: 30,7243,99.92
Unnamed: 29,7237,99.83
Unnamed: 27,7231,99.75
Unnamed: 26,7210,99.46
ean,5706,78.71
manufacturer,4014,55.37
prices.shipping,2972,41.00


In [50]:
duplicate_rows = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_rows)

Number of duplicate rows: 0


### Duplicate Check

The dataset contains **0 exact duplicate rows**, so no records were removed for duplication.

Repeated product IDs are not automatically treated as duplicates because the same product may appear across different merchants, prices, or observation dates.

# copy of dataframe 

In [51]:
df_clean = df.copy()

In [86]:
df_clean = df_clean.drop(
    columns=[
        "ean",
        "manufacturer",
        "Unnamed: 26",
        "Unnamed: 27",
        "Unnamed: 28",
        "Unnamed: 29",
        "Unnamed: 30"
    ],
    errors="ignore"
)

print("Shape after removing unnecessary columns:", df_clean.shape)

Shape after removing unnecessary columns: (7235, 33)


In [53]:
df_clean.columns.tolist()
df_clean.shape

(7249, 24)

In [54]:
df_clean["prices.shipping"].value_counts(dropna=False).head(20)

prices.shipping
NaN                                                 2972
Free Expedited Shipping for most orders over $49    1449
FREE                                                 737
Value                                                697
Standard                                             653
Expedited                                            411
Freight                                              101
Free Delivery                                         71
Free Expedited Shipping                               46
Free Shippingon orders 35 and up                      16
Free Shipping on orders 35 and up                      9
Free Shipping                                          6
Free Shipping for this Item                            6
Free Standard Shipping on Orders Over $49              3
USD 6.65                                               3
USD 10.00                                              2
Shipping Charges Apply                                 2
USD 196.92     

### Categorical Column Inspection

Check unique values in important categorical columns to identify inconsistent spelling, capitalization, formatting, and unexpected values before cleaning.

In [55]:
df_clean["prices.condition"].value_counts(dropna=False)

prices.condition
New                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    

In [56]:
df_clean["prices.availability"].value_counts(dropna=False)

prices.availability
In Stock           3172
Yes                2136
yes                 893
TRUE                663
Out Of Stock        115
Special Order       109
More on the Way      91
undefined            40
sold                 22
No                    4
FALSE                 1
Retired               1
32 available          1
7 available           1
Name: count, dtype: int64

In [57]:
df_clean["prices.currency"].value_counts(dropna=False)

prices.currency
USD    7248
CAD       1
Name: count, dtype: int64

In [58]:
df_clean["prices.isSale"].value_counts(dropna=False)

prices.isSale
False    5508
True     1741
Name: count, dtype: int64

In [59]:
df_clean["primaryCategories"].value_counts(dropna=False)

primaryCategories
Electronics              7196
 Apple CarPlay             27
 Intel Celeron             12
Electronics,Furniture       8
 Siri Eyes Free             6
Name: count, dtype: int64

In [60]:
#df_clean["brand"].value_counts(dropna=False).head(20)
print("Number of unique brands:", df_clean["brand"].nunique())

df_clean["brand"].value_counts(dropna=False).head(30)

Number of unique brands: 257


brand
Sony               785
Samsung            744
Apple              248
Yamaha             240
Pioneer            176
LG                 175
Logitech           116
Lenovo             110
WD                 108
CORSAIR            103
Sennheiser          99
Alpine              93
Elite Screens       93
Corsair             92
Kenwood             91
SanDisk             91
Razer               86
Seagate             86
ASUS                81
Netgear             71
Panasonic           70
Onkyo               68
Denon               64
Lowepro             62
Alienware           62
JBL                 61
Dell                58
Canon               56
TP-Link             55
Western Digital     51
Name: count, dtype: int64

In [61]:
#df_clean["prices.merchant"].value_counts(dropna=False).head(20)
print("Number of unique merchants:", df_clean["prices.merchant"].nunique())

df_clean["prices.merchant"].value_counts(dropna=False).head(30)

Number of unique merchants: 693


prices.merchant
Bestbuy.com                             2806
bhphotovideo.com                        1509
Walmart.com                              664
Beach Camera                             201
AMI Ventures Inc                          63
buydig                                    53
Focus Camera                              48
gear4less                                 43
echo-and-optics                           40
Beach Audio Inc                           39
Best Buy                                  38
UnbeatableSale                            33
DealClock                                 32
World Wide Stereo                         28
electronic_express                        28
BuyVPC                                    27
Electronics Expo (Authorized Dealer)      26
Car Audio Closeout                        24
Video & Audio Center                      19
OneCall                                   18
Electronic Express                        17
wwstereo                               

In [62]:
df_clean[["prices.amountMin", "prices.amountMax"]].describe()

,prices.amountMin,prices.amountMax
count,7249.000000,7249.000000
mean,463.963005,495.530610
std,680.504507,763.599387
min,1.000000,1.000000
25%,79.950000,79.990000
50%,189.990000,198.990000
75%,479.990000,494.990000
max,5999.990000,6999.990000


In [63]:
(df_clean["prices.amountMin"] > df_clean["prices.amountMax"]).sum()

np.int64(0)

In [64]:
df_clean["prices.dateSeen"].head(10)
df_clean["prices.dateSeen"].str.contains(",").sum()

np.int64(3575)

In [65]:
df_clean["categories"].head(10)

0    Audio & Video Accessories,TV Mounts,TV Accesso...
1    Stereos,Portable Bluetooth Speakers,TV, Video ...
2    Stereos,Portable Bluetooth Speakers,TV, Video ...
3    Stereos,Portable Bluetooth Speakers,TV, Video ...
4    Stereos,Portable Bluetooth Speakers,TV, Video ...
5    Stereos,Portable Bluetooth Speakers,TV, Video ...
6    Stereos,Portable Bluetooth Speakers,TV, Video ...
7    Stereos,Portable Bluetooth Speakers,TV, Video ...
8    Stereos,Portable Bluetooth Speakers,TV, Video ...
9    Stereos,Portable Bluetooth Speakers,TV, Video ...
Name: categories, dtype: str

In [66]:
first_date = df_clean["prices.dateSeen"].str.split(",").str[0]

pd.to_datetime(first_date).agg(["min", "max"])

min   2014-08-30 16:00:00+00:00
max   2018-07-25 22:00:00+00:00
Name: prices.dateSeen, dtype: datetime64[us, UTC]

# 1 prices.condition Cleaning

In [67]:
df_clean["condition_clean"] = (
    df_clean["prices.condition"]
    .str.strip()
    .str.lower()
)
df_clean["condition_clean"].value_counts(dropna=False)
def clean_condition(value):
    if pd.isna(value):
        return "unknown"
    
    if "refurbished" in value:
        return "refurbished"
    
    if value == "pre-owned" or value == "used":
        return "used"
    
    if value.startswith("new"):
        return "new"
    
    return "other"

df_clean["condition_clean"] = df_clean["condition_clean"].apply(clean_condition)
df_clean["condition_clean"].value_counts()

condition_clean
new            7019
used            146
refurbished      83
other             1
Name: count, dtype: int64

# 2 prices.availability cleaning

In [68]:
df_clean["availability_clean"] = (
    df_clean["prices.availability"]
    .str.strip()
    .str.lower()
)
df_clean["availability_clean"].value_counts(dropna=False)

def clean_availability(value):
    if value in ["in stock", "yes", "true", "32 available", "7 available"]:
        return "available"
    
    elif value in ["out of stock", "no", "false", "sold", "retired"]:
        return "unavailable"
    
    elif value in ["special order", "more on the way"]:
        return "limited/pending"
    
    else:
        return "unknown"

df_clean["availability_clean"] = (
    df_clean["availability_clean"].apply(clean_availability)
)
df_clean["availability_clean"].value_counts()

availability_clean
available          6866
limited/pending     200
unavailable         143
unknown              40
Name: count, dtype: int64

# 3 Brand cleaning

In [69]:
df_clean["brand_clean"] = (
    df_clean["brand"]
    .str.strip()
    .str.title()
)
df_clean["brand_clean"] = df_clean["brand_clean"].replace(
    {"Wd": "Western Digital"}
)
print("Original unique brands:", df_clean["brand"].nunique())
print("Clean unique brands:", df_clean["brand_clean"].nunique())


Original unique brands: 257
Clean unique brands: 237


# 4 Merchant cleaning

In [70]:
df_clean["merchant_clean"] = (
    df_clean["prices.merchant"]
    .str.strip()
    .str.lower()
)
print("Original unique merchants:", df_clean["prices.merchant"].nunique())
print("Clean unique merchants:", df_clean["merchant_clean"].nunique())

Original unique merchants: 693
Clean unique merchants: 681


# 5 Shipping cleaning

In [71]:
df_clean["shipping_clean"] = (
    df_clean["prices.shipping"]
    .str.strip()
    .str.lower()
)
df_clean["shipping_clean"].value_counts(dropna=False).head(20)

#category defining 
def categorize_shipping(value):
    if pd.isna(value):
        return "unknown"
    
    if "free" in value:
        if "over" in value or "orders 35" in value or "orders $35" in value:
            return "conditional free"
        else:
            return "free"
    
    return "paid/other"

df_clean["shipping_type"] = (
    df_clean["shipping_clean"].apply(categorize_shipping)
)
df_clean["shipping_type"].value_counts()

shipping_type
unknown             2972
paid/other          1931
conditional free    1477
free                 869
Name: count, dtype: int64

# 6 Currency
# 1 out of 7,249 rows (0.014%), converting it to USD would add unnecessary complexity and introduce an exchange-rate assumption. The cleanest methodology is to exclude the CAD row and analyze USD listings only.

In [72]:
df_clean.loc[
    df_clean["prices.currency"] == "CAD",
    ["name", "brand", "prices.amountMin", "prices.amountMax",
     "prices.merchant", "prices.currency"]
]
df_clean = df_clean[
    df_clean["prices.currency"] == "USD"
].copy()
print("Rows after currency cleaning:", len(df_clean))
print(df_clean["prices.currency"].value_counts())

Rows after currency cleaning: 7248
prices.currency
USD    7248
Name: count, dtype: int64


# 7 clean the date
# For this project, we'll create a new column containing the most recent date seen for each listing. We will keep the original prices.dateSeen untouched.

In [73]:
df_clean["date_seen_clean"] = df_clean["prices.dateSeen"].apply(
    lambda x: max(pd.to_datetime(x.split(",")))
)

df_clean["date_seen_clean"].head(10)

print("Missing cleaned dates:", df_clean["date_seen_clean"].isna().sum())
print("Earliest date:", df_clean["date_seen_clean"].min())
print("Latest date:", df_clean["date_seen_clean"].max())

Missing cleaned dates: 0
Earliest date: 2014-08-30 16:00:00+00:00
Latest date: 2018-07-25 23:00:00+00:00


# one usable price column

In [74]:
df_clean["price"] = (
    df_clean["prices.amountMin"] + df_clean["prices.amountMax"]
) / 2
df_clean["price"].describe()

df_clean[
    ["name", "brand_clean", "price", "prices.merchant"]
].sort_values("price").head(10)
print("Under $5:", (df_clean["price"] < 5).sum())
print("Under $10:", (df_clean["price"] < 10).sum())
print("Under $20:", (df_clean["price"] < 20).sum())
print("Under $50:", (df_clean["price"] < 50).sum())

df_clean.loc[
    df_clean["price"] < 5,
    ["name", "brand_clean", "price", "prices.amountMin",
     "prices.amountMax", "merchant_clean"]
].sort_values("price")

df_clean = df_clean[df_clean["price"] >= 5].copy()
print("Rows remaining:", len(df_clean))
print("Minimum price:", df_clean["price"].min())
df_clean[
    ["name", "brand_clean", "price", "prices.amountMin",
     "prices.amountMax", "merchant_clean"]
].sort_values("price", ascending=False).head(10)

Under $5: 13
Under $10: 37
Under $20: 115
Under $50: 923
Rows remaining: 7235
Minimum price: 6.45


,name,brand_clean,price,prices.amountMin,prices.amountMax,merchant_clean
5900,Samsung - 65 Class - LED - MU6290 Series - 216...,Samsung,6498.990,5997.99,6999.99,walmart.com
5890,Samsung - 65 Class - LED - MU6290 Series - 216...,Samsung,5999.990,5999.99,5999.99,bestbuy.com
3638,XBR-X850E-Series 75-Class HDR UHD Smart LED TV,Sony,4999.990,4999.99,4999.99,bestbuy.com
2286,Sony - 55 Class - OLED - A1E Series - 2160p - ...,Sony,4999.990,4999.99,4999.99,bestbuy.com
2287,Sony - 55 Class - OLED - A1E Series - 2160p - ...,Sony,4647.995,4296.00,4999.99,beach camera
2290,Sony - 55 Class - OLED - A1E Series - 2160p - ...,Sony,4547.995,4096.00,4999.99,beach camera
5740,LG Electronics OLED65C7P 65-Inch 4K Ultra HD S...,Lg,4496.990,4496.99,4496.99,bhphotovideo.com
1422,Razer - Blade Pro 17.3 4K Ultra HD Touch-Scree...,Razer,4399.990,4399.99,4399.99,bestbuy.com
1424,Razer - Blade Pro 17.3 4K Ultra HD Touch-Scree...,Razer,4399.990,4399.99,4399.99,bestbuy.com
3454,Refurbished Samsung Curved 65 4K (2160P) Smart...,Samsung,4348.990,3497.99,5199.99,walmart.com


# final check

In [75]:
df_clean[
    [
        "price",
        "brand_clean",
        "merchant_clean",
        "condition_clean",
        "availability_clean",
        "shipping_type",
        "date_seen_clean"
    ]
].isna().sum()

price                 0
brand_clean           0
merchant_clean        0
condition_clean       0
availability_clean    0
shipping_type         0
date_seen_clean       0
dtype: int64

# inspect category structure

In [76]:
df_clean["categories"].head(10).tolist()
category_counts = (
    df_clean["categories"]
    .str.split(",")
    .explode()
    .str.strip()
    .value_counts()
)

category_counts.head(30)
print("Total listings:", len(df_clean))
print("Unique products:", df_clean["id"].nunique())

Total listings: 7235
Unique products: 835


In [77]:
products = (
    df_clean[
        ["id", "name", "brand_clean", "categories"]
    ]
    .drop_duplicates(subset="id")
    .copy()
)

print("Product rows:", len(products))
products.head()

products["name"].head(30).tolist()

Product rows: 835


["Sanus VLF410B1 10-Inch Super Slim Full-Motion Mount for 37 - 84 Inches TV's",
 'Boytone - 2500W 2.1-Ch. Home Theater System - Black Diamond',
 'DENAQ - AC Adapter for TOSHIBA SATELLITE 1700 1710 1715 1730 1735 1750 1755 1955 3000 3005 A100 M30X M35X - Black',
 'DreamWave - Tremor Portable Bluetooth Speaker - Green,Black',
 'NS-SP1800BL 5.1-Channel Home Theater System (Black)',
 'Universal Remote Control - 48-Device Universal Remote - Black',
 'Acoustimass 6 Series V Home Theater Speaker System (Black)',
 'Samsung - 850 PRO 512GB Internal SATA III Solid State Drive for Laptops',
 'Corsair Vengeance LPX 16GB (2x8GB) DDR4 DRAM 3000MHz C15 Desktop Memory Kit - Black (CMK16GX4M2B3000C15)',
 'Outdoor Tech Buckshot Pro Bluetooth Speaker',
 'Motorola Wi-Fi Pet Video Camera',
 'Details About Samsung Gear Iconx 2018 Edition Cordfree Fitness Earbuds Black (us Version)',
 '2TB Red 5400 rpm SATA III 3.5 Internal NAS HDD',
 'Details About Panamax Mr4000 8outlets Surge Protector Home Theater Power 

In [78]:
print("Unique products:", len(products))
print("Unique category combinations:", products["categories"].nunique())

Unique products: 835
Unique category combinations: 803


In [79]:
keywords = [
    "computer", "laptop", "tablet",
    "tv", "television",
    "audio", "speaker", "headphone",
    "camera",
    "mobile", "phone",
    "car", "auto"
]

for word in keywords:
    count = products["categories"].str.contains(
        word, case=False, na=False
    ).sum()
    print(f"{word}: {count}")


    

computer: 288
laptop: 67
tablet: 244
tv: 285
television: 91
audio: 475
speaker: 193
headphone: 108
camera: 93
mobile: 99
phone: 160
car: 175
auto: 76


In [80]:
def assign_category(row):
    text = (
        str(row["name"]) + " " + str(row["categories"])
    ).lower()

    if any(word in text for word in [
        "car electronics", "car audio", "auto electronics",
        "automotive", "car speaker"
    ]):
        return "Automotive Electronics"

    elif any(word in text for word in [
        "camera", "camcorder", "photography"
    ]):
        return "Cameras & Photography"

    elif any(word in text for word in [
        "phone", "mobile", "smartphone", "iphone"
    ]):
        return "Mobile & Phones"

    elif any(word in text for word in [
        "computer", "laptop", "tablet", "desktop",
        "ssd", "hard drive", "memory"
    ]):
        return "Computers & Tablets"

    elif any(word in text for word in [
        "tv", "television", "home theater", "projector"
    ]):
        return "TV & Home Theater"

    elif any(word in text for word in [
        "audio", "speaker", "headphone", "earbud", "stereo"
    ]):
        return "Audio"

    else:
        return "Other Electronics"


products["category_clean"] = products.apply(assign_category, axis=1)
products["category_clean"].value_counts()

category_clean
Computers & Tablets       236
TV & Home Theater         204
Mobile & Phones           157
Automotive Electronics     92
Cameras & Photography      87
Audio                      58
Other Electronics           1
Name: count, dtype: int64

In [81]:
products[
    products["category_clean"] == "Other Electronics"
][["name", "brand_clean", "categories"]]

,name,brand_clean,categories
5729,Energizer - Recharge Rechargeable AA Batteries...,Energizer,"Electrical,Household Batteries,AA Batteries,Co..."


In [82]:
df_clean = df_clean.merge(
    products[["id", "category_clean"]],
    on="id",
    how="left"
)
print("Rows:", len(df_clean))

print("Missing categories:", df_clean["category_clean"].isna().sum())
df_clean["category_clean"].value_counts()

Rows: 7235
Missing categories: 0


category_clean
Computers & Tablets       2199
TV & Home Theater         1913
Mobile & Phones           1145
Cameras & Photography      881
Automotive Electronics     796
Audio                      299
Other Electronics            2
Name: count, dtype: int64

In [83]:
df_clean = df_clean.drop(
    columns=["category_clean_x", "category_clean_y"],
    errors="ignore"
)

print(df_clean.shape)

(7235, 33)


# save clean data set

In [87]:
df_clean.to_csv(
    "../data/processed/electronics_pricing_clean.csv",
    index=False
)

print("Cleaned dataset saved!")
print("Final rows:", len(df_clean))
print("Final columns:", len(df_clean.columns))

Cleaned dataset saved!
Final rows: 7235
Final columns: 33


# final validation

In [85]:
print("FINAL DATA QUALITY CHECK")
print("-" * 30)

print("Rows:", len(df_clean))
print("Columns:", len(df_clean.columns))
print("Unique products:", df_clean["id"].nunique())

print("\nMissing values in analysis columns:")
print(
    df_clean[
        [
            "price",
            "brand_clean",
            "merchant_clean",
            "condition_clean",
            "availability_clean",
            "shipping_type",
            "date_seen_clean",
            "category_clean"
        ]
    ].isna().sum()
)

print("\nExact duplicate rows:", df_clean.duplicated().sum())

print("\nPrice range:")
print("Minimum:", df_clean["price"].min())
print("Maximum:", df_clean["price"].max())

print("\nCurrency:")
print(df_clean["prices.currency"].value_counts())

FINAL DATA QUALITY CHECK
------------------------------
Rows: 7235
Columns: 33
Unique products: 835

Missing values in analysis columns:
price                 0
brand_clean           0
merchant_clean        0
condition_clean       0
availability_clean    0
shipping_type         0
date_seen_clean       0
category_clean        0
dtype: int64

Exact duplicate rows: 0

Price range:
Minimum: 6.45
Maximum: 6498.99

Currency:
prices.currency
USD    7235
Name: count, dtype: int64


## Data Cleaning Summary

The raw electronics pricing dataset was assessed and cleaned before analysis.

### Cleaning Steps
- Removed columns with extremely high missing-value rates that were not useful for the analysis.
- Checked for exact duplicate records; no exact duplicates were found.
- Standardized product condition into broader groups: New, Used, Refurbished, and Other.
- Standardized product availability into Available, Unavailable, Limited/Pending, and Unknown.
- Standardized brand and merchant text to reduce inconsistencies caused by capitalization and whitespace.
- Grouped shipping information into Free, Conditional Free, Paid/Other, and Unknown.
- Restricted the analysis to USD listings. One CAD listing was excluded rather than converted because it represented only one record.
- Parsed `prices.dateSeen` and selected the most recent timestamp when multiple observation dates were stored in a single record.
- Created a representative `price` variable using the midpoint of `prices.amountMin` and `prices.amountMax`.
- Investigated unusually low prices and removed 13 listings below $5 after confirming that they represented implausible prices for major electronics products.
- Investigated high-priced observations and retained them because they represented plausible premium electronics such as televisions and laptops.
- Created broader business-friendly product categories using product names and existing category information.
- Preserved repeated product listings because the same product can appear across different merchants, prices, and observations.

### Final Dataset
- Listings: 7,235
- Unique products: 835
- Core analysis columns with missing values: 0
- Exact duplicate rows: 0
- Currency: USD
- Price range: $6.45–$6,498.99

The cleaned dataset was saved as `data/processed/electronics_pricing_clean.csv` for exploratory data analysis.